### PR TITLE
[ADD][14.0] stock_unique_order_per_location + shopfloor_single_product_transfer_unique_order_at_location

### PR DESCRIPTION
**Pull request by TDu**
This PR is based on the scenario single product transfer
* Backend : https://github.com/OCA/wms/pull/513
* Frontend: https://github.com/OCA/wms/pull/520

ref.: rau-131

### CODE REVIEW

**Comment by TDu** [comment]
Error in the tests are probably not related to this pr.
Anyway we still depends on the `shopfloor_single_product_transfer` PR #513

File: shopfloor_single_product_transfer_same_next_picking/services/single_product_transfer.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+                ),
+            ]
+        )
+        if not lines:
+            return res
+        # From the move line to work on get the picking for its destination moves
+        same_picking = move_line.move_id.move_dest_ids.picking_id
+        # In the destination there should be only moves from that picking.
+        if all(move in same_picking.move_lines for move in lines.move_id):
+            return res
```
Comment (jbaudoux) [inline shopfloor_single_product_transfer_same_next_picking/services/single_product_transfer.py]: This should be in a module in the backend that will ensure the constrain on action done. The integration in shopfloor should be on top of that one to anticipate the error but where the logic is

**Review by jbaudoux — CHANGES_REQUESTED** [review]

**Review by jbaudoux — CHANGES_REQUESTED** [review]
We also lack the configuration on the picking type to enable the feature

**Comment by TDu** [comment]
@jbaudoux I refactored and added the option on operation type. The error in the tests seems unrelated.

File: stock_location_donot_mix_order/models/stock_picking_type.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from odoo import fields, models
+
+
+class StockPickingType(models.Model):
+
+    _inherit = "stock.picking.type"
+
+    same_next_picking = fields.Boolean(
+        string="Only same order at desination",
```
Comment (jbaudoux) [inline stock_location_donot_mix_order/models/stock_picking_type.py]:
```suggestion
        string="Do not mix sources at destination",
```

File: stock_location_donot_mix_order/models/stock_picking_type.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+        help="When checked, only products ordered for the same sales order will be"
+        "allowed in the destination location.",
```
Comment (jbaudoux) [inline stock_location_donot_mix_order/models/stock_picking_type.py]:
```suggestion
        help="When checked, the destination location must either be empty or contain stock reserved only by the next picking. This ensures that you do not mix different sales order on the destination location.",
```

File: stock_location_donot_mix_order/models/stock_move_line.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+class StockMoveLine(models.Model):
+    _inherit = "stock.move.line"
+
+    def _action_done(self):
+        self._check_same_order_at_destination()
+        return super()._action_done()
+
+    def _check_same_order_at_destination(self, location=None):
+        move_lines = self.filtered(
+            lambda line: line.move_id.picking_id.picking_type_id.same_next_picking
```
Comment (jbaudoux) [inline stock_location_donot_mix_order/models/stock_move_line.py]:
```suggestion
            lambda line: line.move_id.picking_type_id.same_next_picking
```

Comment (jbaudoux) [reply stock_location_donot_mix_order/models/stock_move_line.py]: prevent looping twice over the move lines by including this in the for loop.
```
If not line.move_id.picking_type_id.same_next_picking:
  continue
```

File: stock_location_donot_mix_order/models/stock_move_line.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                            "assigned",
+                            "partially_available",
+                        ),
+                    ),
+                ]
+            )
+            if not lines:
+                continue
+            # From the move line to work on get the picking for its destination moves
+            same_picking = line.move_id.move_dest_ids.picking_id
```
Comment (jbaudoux) [inline stock_location_donot_mix_order/models/stock_move_line.py]: name it properly
```suggestion
            next_picking = line.move_id.move_dest_ids.picking_id
```

File: stock_location_donot_mix_order/models/stock_move_line.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            # In the destination there should be only moves from that picking.
+            if all(move in same_picking.move_lines for move in lines.move_id):
```
Comment (jbaudoux) [inline stock_location_donot_mix_order/models/stock_move_line.py]: You can include the condition in the search to not loop in python. If we find a move line that is not one of the next picking, then there is an error:
```
if search([..., ("id", "not in", next_picking.move_line_ids.ids)], limit=1):
  raise Error
```

File: stock_location_donot_mix_order/models/stock_move_line.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    def _action_done(self):
+        self._check_same_order_at_destination()
+        return super()._action_done()
+
+    def _check_same_order_at_destination(self, location=None):
+        move_lines = self.filtered(
+            lambda line: line.move_id.picking_id.picking_type_id.same_next_picking
+        )
+        for line in move_lines:
+            destination_location = location or line.location_dest_id
```
Comment (jbaudoux) [inline stock_location_donot_mix_order/models/stock_move_line.py]: You could group the lines by picking to not make a search for each line
```suggestion
        for picking, lines in move_lines_by_picking:
            destination_locations = location or lines.location_dest_id
```

File: stock_location_donot_mix_order/models/stock_move_line.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            if not destination_location.quant_ids:
+                continue
```
Comment (jbaudoux) [inline stock_location_donot_mix_order/models/stock_move_line.py]: I think this check is useless as it is already covered by the next one.
What would be useful is to check that there are no unreserved quants on the location
```
if any(q.available_quantity for q in destination_locations.quant_ids):
  raise Error
```

File: stock_location_donot_mix_order/models/stock_picking_type.py [hunk 9]
```diff
@@ -0,0 +1,14 @@
+# rCopyright 2023 Camptocamp SA
```
Comment (jbaudoux) [inline stock_location_donot_mix_order/models/stock_picking_type.py]: typo
```suggestion
# Copyright 2023 Camptocamp SA
```

File: stock_location_donot_mix_order/models/stock_move_line.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+from odoo import _, models
+from odoo.exceptions import UserError
+
+
+class StockMoveLine(models.Model):
+    _inherit = "stock.move.line"
+
+    def _action_done(self):
+        self._check_same_order_at_destination()
+        return super()._action_done()
```
Comment (jbaudoux) [inline stock_location_donot_mix_order/models/stock_move_line.py]: It is safer to do the check after the moves are done.
If you have 2 move lines of 2 pickings going to the same destination, if will fail only if you check it after the move are done, not before.
Won't happen, but for correctness it is better.

File: stock_location_donot_mix_order/__manifest__.py [hunk 11]
```diff
@@ -0,0 +1,13 @@
+{
+    "name": "Stock Location Do Not Mix Order",
+    "summary": "Prevent to mix sales order on same consolidation location.",
```
Comment (jbaudoux) [inline stock_location_donot_mix_order/__manifest__.py]: After rethinking on the name, I don't like that it starts with stock_location as it's like it's adding a feature on the location while it is a feature on a picking.
Maybe `stock_unique_order_per_location` ?
```suggestion
    "name": "Stock Unique Order Per Location",
    "summary": "Prevent to mix different sources on a same location.",
```

File: stock_location_donot_mix_order/__manifest__.py [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+    "version": "14.0.1.0.0",
+    "category": "Inventory",
+    "website": "https://github.com/OCA/wms",
+    "author": "Camptocamp, Odoo Community Association (OCA)",
```
Comment (jbaudoux) [inline stock_location_donot_mix_order/__manifest__.py]: :)
```suggestion
    "author": "Camptocamp, BCIM, Odoo Community Association (OCA)",
```

**Review by jbaudoux — CHANGES_REQUESTED** [review]

File: stock_location_donot_mix_order/models/stock_move_line.py (hunk 8 above)
Comment (TDu) [reply stock_location_donot_mix_order/models/stock_move_line.py]: I changed this to 
```
if not any(quant.available_quantity for quant in destination_location.quant_ids):
    continue
```

File: stock_location_donot_mix_order/__manifest__.py (hunk 11 above)
Comment (TDu) [reply stock_location_donot_mix_order/__manifest__.py]: I also renamed the related shopfloor module

File: stock_unique_order_per_location/models/stock_move_line.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    def _check_same_order_at_destination(self, location=None):
+        move_lines = self.filtered(
+            lambda line: line.move_id.picking_type_id.same_next_picking
+        )
+        for line in move_lines:
+            destination_location = location or line.location_dest_id
+            if not any(
+                quant.available_quantity for quant in destination_location.quant_ids
+            ):
+                continue
```
Comment (jbaudoux) [inline stock_unique_order_per_location/models/stock_move_line.py]: This is not correct. Either there are quants available and this is an error.
Or there is no quant available and then you need to perform the second check on the move lines

File: stock_unique_order_per_location/models/stock_move_line.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        res = super()._action_done()
+        self._check_same_order_at_destination()
+        return res
+
+    def _check_same_order_at_destination(self, location=None):
+        move_lines = self.filtered(
+            lambda line: line.move_id.picking_type_id.same_next_picking
+        )
+        for line in move_lines:
+            destination_location = location or line.location_dest_id
```
Comment (jbaudoux) [inline stock_unique_order_per_location/models/stock_move_line.py]: Don't loop twice over the move lines and don't make a search query for each move line. You should do this check at picking level.
```
for picking_id, move_lines in odoo.tools.groupby(self, lambda line: line.picking_id):
  if not picking_id.picking_type_id.same_next_picking:
    continue
  move_lines = self.browse().union(*move_lines)
  destination_locations = location or move_lines.location_dest_id
  ...
```

**Review by jbaudoux — CHANGES_REQUESTED** [review]

File: stock_location_donot_mix_order/models/stock_move_line.py (hunk 7 above)
Comment (jbaudoux) [reply stock_location_donot_mix_order/models/stock_move_line.py]: comment added back on the new code, marking this one as resolved

File: stock_unique_order_per_location/models/stock_move_line.py (hunk 14 above)
Comment (TDu) [reply stock_unique_order_per_location/models/stock_move_line.py]: done

File: stock_unique_order_per_location/models/stock_move_line.py (hunk 13 above)
Comment (TDu) [reply stock_unique_order_per_location/models/stock_move_line.py]: fixed

**Comment by TDu** [comment]
Error in the tests do not seem to be related to the new modules.
I do not understand it either https://github.com/OCA/wms/actions/runs/4499025827/jobs/7916362523?pr=603#step:8:41

File: stock_unique_order_per_location/__manifest__.py [hunk 15]
```diff
@@ -0,0 +1,13 @@
+{
+    "name": "Stock Unique Order Per Location",
+    "summary": "Prevent to mix sales order on same consolidation location.",
```
Comment (sebalix) [inline stock_unique_order_per_location/__manifest__.py:3]: Could be used for any kind of transfer, like internal ones, you have no strict dependency on sale order here.
```suggestion
    "summary": "Prevent to mix transfers on same consolidation location.",
```

File: stock_unique_order_per_location/models/stock_move_line.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        # Will not have happen yet (in stock.move _action_done)
+        # The quantity will still be available in the quants
+        self._check_same_order_at_destination()
+        return super()._action_done()
+
+    @api.model
+    def _raise_error_unique_order_per_location(self, locations):
+        raise UserError(
+            _(
+                "The location(s) {} can only accept goods for the same sales order.".format(
```
Comment (sebalix) [inline stock_unique_order_per_location/models/stock_move_line.py:23]:
```suggestion
                "The location(s) {} can only accept goods from the same ongoing transfer.".format(
```

File: stock_unique_order_per_location/models/stock_picking_type.py [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+
+class StockPickingType(models.Model):
+
+    _inherit = "stock.picking.type"
+
+    same_next_picking = fields.Boolean(
+        string="Do not mix sources at destination",
+        help="When checked, the destination location must either be empty or "
+        "contain stock reserved only by the next picking. This ensures that you "
+        "do not mix different sales order on the destination location",
```
Comment (sebalix) [inline stock_unique_order_per_location/models/stock_picking_type.py:14]:
```suggestion
        "do not mix different transfers on the destination location.",
```

File: stock_unique_order_per_location/readme/DESCRIPTION.rst [hunk 18]
```diff
@@ -0,0 +1,6 @@
+This module adds an option on operation types.
+The `Only Same Order At Destination` option when checked, will only
+allow on a location products ordered for the same sales order.
```
Comment (sebalix) [inline stock_unique_order_per_location/readme/DESCRIPTION.rst:3]:
```suggestion
The `Do not mix sources at destination` option when checked, will only
allow on a location products coming from the same transfer.
```

File: shopfloor_single_product_transfer_unique_order_at_location/__manifest__.py [hunk 19]
```diff
@@ -0,0 +1,12 @@
+{
+    "name": "Shopfloor Single Product Transfer Unique Order Per Location",
+    "summary": "Prevent to mix sales order on same consolidation location.",
```
Comment (sebalix) [inline shopfloor_single_product_transfer_unique_order_at_location/__manifest__.py:3]:
```suggestion
    "name": "Shopfloor Single Product Transfer Unique Order Per Location",
    "summary": "Prevent to mix ongoing transfers on same consolidation location.",
```

File: stock_unique_order_per_location/models/stock_move_line.py [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+                            "partially_available",
+                        ),
+                    ),
+                    # Search for lines at location not in the next picking
+                    (
+                        "move_id.id",
+                        "not in",
+                        move_lines.move_id.move_dest_ids.picking_id.move_lines.ids,
+                    ),
+                ]
```
Comment (sebalix) [inline stock_unique_order_per_location/models/stock_move_line.py]: Domain to put in a separate method to ease override

Comment (TDu) [reply stock_unique_order_per_location/models/stock_move_line.py]: done

**Comment by TDu** [comment]
For the change in the messages and documentation, I would like @jbaudoux to have his say.

File: shopfloor_single_product_transfer_unique_order_at_location/services/single_product_transfer.py [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+from odoo.exceptions import UserError
+
+from odoo.addons.component.core import Component
+
+
+class ShopfloorSingleProductTransfer(Component):
+    _inherit = "shopfloor.single.product.transfer"
+
+    def _set_quantity__check_location(self, location, move_line, confirmation=False):
+        res = super()._set_quantity__check_location(location, move_line, confirmation)
```
Comment (mmequignon) [inline shopfloor_single_product_transfer_unique_order_at_location/services/single_product_transfer.py]:
```suggestion
    def _set_quantity__check_location(self, move-line, location=False, confirmation=False):
        res = super()._set_quantity__check_location(move_line, location=location confirmation)
```

**Review by mmequignon — CHANGES_REQUESTED** [review]

File: shopfloor_single_product_transfer_unique_order_at_location/services/single_product_transfer.py (hunk 21 above)
Comment (jbaudoux) [reply shopfloor_single_product_transfer_unique_order_at_location/services/single_product_transfer.py]: due to the change proposed in https://github.com/OCA/wms/pull/640

**Comment by jbaudoux** [comment]
> For the change in the messages and documentation, I would like @jbaudoux to have his say.

I would not change the messages and keep sales order. That's currently the unique usage

File: shopfloor_single_product_transfer_unique_order_at_location/services/single_product_transfer.py (hunk 21 above)
Comment (jbaudoux) [reply shopfloor_single_product_transfer_unique_order_at_location/services/single_product_transfer.py]: @TDu Can you rebase and fix this ?

**Review by jbaudoux — APPROVED** [review]

**Comment by jbaudoux** [comment]
@mmequignon Can you update your review?

**Review by JuMiSanAr — APPROVED** [review]

**Pull request merged**